arangod/Aql/Ast.cpp

```diff
@@ -1430,107 +1427,88 @@ void Ast::injectBindParameters(BindParameters& parameters) {
 
       TRI_ASSERT(!param.empty());
 
-      switch (node->getDataSourceType()) {
-        case AstNode::DataSourceType::Collection: {
-          // bound collection parameter
+      if ('@' == param[0]) {
+          // bound data source parameter
           TRI_ASSERT(value.isString());
 
-          // check if the collection was used in a data-modification query
-          bool isWriteCollection = false;
-
-          for (auto const& it : _writeCollections) {
-            if (it->type == NODE_TYPE_PARAMETER && StringRef(param) == StringRef(it->getStringValue(), it->getStringLength())) {
-              isWriteCollection = true;
-              break;
-            }
-          }
-
-          // turn node into a collection node
           char const* name = nullptr;
           VPackValueLength length;
           char const* stringValue = value.getString(length);
 
-          if (length > 0 && stringValue[0] >= '0' && stringValue[0] <= '9') {
-            // emergency translation of collection id to name
-            arangodb::CollectionNameResolver resolver(_query->vocbase());
-            std::string collectionName = resolver.getCollectionNameCluster(basics::StringUtils::uint64(stringValue, length));
-            if (collectionName.empty()) {
-              THROW_ARANGO_EXCEPTION_PARAMS(TRI_ERROR_ARANGO_COLLECTION_NOT_FOUND,
-                                            value.copyString().c_str());
-            }
+          // FIXME use external resolver
+          arangodb::CollectionNameResolver resolver(_query->vocbase());
```

Comment:
use `_query->trx()->resolver()`

Reply:
@graetzer thank you for your suggestion but unfortunately there may be no transaction at all at this point of query execution:
```
ExecutionPlan* Query::preparePlan() {
  LOG_TOPIC(DEBUG, Logger::QUERIES) << TRI_microtime() - _startTime << " "
                                    << "Query::prepare"
                                    << " this: " << (uintptr_t) this;
  std::unique_ptr<ExecutionPlan> plan;

  if (!_queryString.empty()) {
    Parser parser(this);
    
    parser.parse(false);
    // put in bind parameters
    parser.ast()->injectBindParameters(_bindParameters);
  }

  TRI_ASSERT(_trx == nullptr);
  
  // TODO: Remove once we have cluster wide transactions
  std::unordered_set<std::string> inaccessibleCollections;
#ifdef USE_ENTERPRISE
  if (_queryOptions.transactionOptions.skipInaccessibleCollections) {
    inaccessibleCollections = _queryOptions.inaccessibleCollections;
  }
#endif
  
  std::unique_ptr<AqlTransaction> trx(AqlTransaction::create(
                     createTransactionContext(), _collections.collections(),
                     _queryOptions.transactionOptions,
                     _part == PART_MAIN, inaccessibleCollections));
  TRI_DEFER(trx.release());
  // create the transaction object, but do not start it yet
  _trx = trx.get();
...
```
That's exactly what I wanted to solve somehow in the next PR

Reply:
@gnusi is right. The code part he changed created a new resolver before already exactly because there may be no transaction yet.